File: ios/Utils/PhotoLibraryProcessor.swift
//
//  PhotoLibraryProcessor.swift
//  Glimpse
//
//  Created by Nitin Issac Joy on 09/07/19.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation
import CoreLocation
import Photos
import RealmSwift

@objc(PhotoLibraryProcessor)
class PhotoLibraryProcessor: NSObject {
  
  static func getPhotosFromLibrary() -> [ClusterModal] {
    let allPhotos = PHAsset.fetchAssets(with: .image, options: PHFetchOptions())
    var arrayOfPHAsset : [ClusterModal] = []
    allPhotos.enumerateObjects({(object: AnyObject!,
      count: Int,
      stop: UnsafeMutablePointer<ObjCBool>) in
      
      if object is PHAsset{
        let asset = object as! PHAsset
        if(asset.location != nil && asset.creationDate != nil) {
          let _el: ClusterModal = ClusterModal()
          _el.image = "ph://" + asset.localIdentifier
          _el.latitude = asset.location?.coordinate.latitude ?? 0
          _el.longitude = asset.location?.coordinate.longitude ?? 0
          _el.timestamp = asset.creationDate?.timeIntervalSince1970 ?? 0
          arrayOfPHAsset.append(_el)
        }
      }
    })
    
    return arrayOfPHAsset
  }
  
  static func GetJPEGFromPHAsset(path: String) -> Data? {
    let asset: PHAsset? = PHAsset.fetchAssets(withLocalIdentifiers: [path], options: .none).firstObject
    if (asset != nil) {
      let retinaSquare = CGSize(width: 1000, height: 1000)
      var thumbnail: UIImage? = nil
      PHImageManager.default().requestImage(for: asset!, targetSize: retinaSquare, contentMode: .aspectFit, options: PHImageRequestOptions(), resultHandler: {(result, info)->Void in
        thumbnail = result!
      })
      if thumbnail != nil {
        return thumbnail!.jpegData(compressionQuality: 100)
      }
      return nil
    }
    return nil
  }
  
  
  static func GenerateTripFromPhotos(clusterData: [ClusterModal], homesForDataClustering: List<HomesForDataClusteringModal>, endTimestamp: Int64) throws -> [TripModal] {
    
    let trips = ClusterProcessor.RunMasterClustering(clusterData: clusterData, homes: homesForDataClustering);
    if (trips.count == 0) {
      throw EngineError.coreEngineError(message: "No trips found")
    }
    
    var tripResult: [TripModal] = [];
    
    var i = 0;
    for trip in trips {
      let _trip = trip.sorted(by: { $0.timestamp < $1.timestamp })
      let _unsortedSteps: [StepModal] = ClusterProcessor.RunStepClustering(trip: _trip);
      let __trip = PhotoLibraryProcessor.PopulateTripModalData(steps: _unsortedSteps.sorted(by: {$0.startTimestamp < $1.startTimestamp }), tripId: TripUtils.GenerateTripId(), homesDataForClustering: homesForDataClustering);
      tripResult.append(__trip);
      i += 1;
    }
    
    tripResult = tripResult.sorted(by: {$0.startDate < $1.startDate})
    
    return tripResult
  }
  
  
  static func PopulateTripModalData(steps: [StepModal], tripId: String, homesDataForClustering: List<HomesForDataClusteringModal>) -> TripModal {
    
    var tripResult: TripModal = TripModal();
    
    var i = 1;
    var _stepsForTrip: [StepModal] = []
    var countries: [String] = []
    var places: [String] = []
    
    _stepsForTrip.append(PhotoLibraryProcessor.GetHomeStepForTimestamp(homesDataForClustering: homesDataForClustering, timestamp: steps[0].startTimestamp, tripId: tripId, stepId: 0));
    
    // Load locations
    for step in steps {
      let result = TripUtils.getLocationFromCoordinates(latitude: step.meanLatitude, longitude: step.meanLongitude)
      step.stepName = result
      step.tripId = tripId;
      step.stepId = i*100;
      
      if (countries.firstIndex(of: result) == nil) {
        countries.append(result)
      }
      
      if (places.firstIndex(of: step.stepName) == nil) {
        places.append(step.stepName)
      }
      
      let _obj = Country()
      _obj.country = result
      tripResult.countryCode.append(_obj)

      // Showing current weather now
      step.temperature = String(TripUtils.getWeatherFromCoordinates(latitude: step.meanLatitude, longitude: step.meanLongitude)) + "ºC"
      step.distanceTravelled = GetDistanceTravelledBetweenSteps(a: steps[i-1], b: step)
      _stepsForTrip.append(step);
      i += 1;
    }
    
    let _stepModal2: StepModal = PhotoLibraryProcessor.GetHomeStepForTimestamp(homesDataForClustering: homesDataForClustering, timestamp: steps[steps.count - 1].endTimestamp, tripId: tripId, stepId: i*100)
    _stepModal2.distanceTravelled = steps[steps.count - 1].distanceTravelled + PhotoLibraryProcessor.GetDistanceTravelledBetweenSteps(a: _stepsForTrip[_stepsForTrip.count - 1], b: _stepModal2)
    _stepsForTrip.append(_stepModal2)
    
    tripResult.tripId = tripId;
    tripResult = PhotoLibraryProcessor.PopulateTripWithSteps(trip: tripResult, steps: _stepsForTrip)
    DatabaseProvider.UpdateDBWithSteps(steps: _stepsForTrip)
    
    return tripResult
  }
  
  static func GetDistanceTravelledBetweenSteps(a: StepModal, b: StepModal) -> Int {
    
    let _p = ClusterModal()
    _p.latitude = a.meanLatitude
    _p.longitude = a.meanLongitude
    
    let _q = ClusterModal()
    _q.latitude = b.meanLatitude
    _q.longitude = b.meanLongitude
    
    return (ClusterProcessor.EarthDistance(p: _p, q: _q))
  }
  
  static func GetHomeStepForTimestamp(homesDataForClustering: List<HomesForDataClusteringModal>, timestamp: TimeInterval, tripId: String, stepId: Int) -> StepModal {
    
    let homeStep = homesDataForClustering[Int(timestamp / 86400) - 1]
    homeStep.timestamp = timestamp
    
    let homeStepCluster = ClusterModal()
    homeStepCluster.latitude = homeStep.latitude
    homeStepCluster.longitude = homeStep.longitude
    homeStepCluster.timestamp = homeStep.timestamp
    
    
    let _stepModal: StepModal = ClusterProcessor.convertClusterToStep(cluster: [homeStepCluster])
    _stepModal.stepName = "Home";
    _stepModal.stepId = stepId;
    _stepModal.tripId = tripId;
    
    return _stepModal
  }
  
  static func PopulateTripWithSteps(trip: TripModal, steps: [StepModal]) -> TripModal {
    trip.daysOfTravel = Int((steps[steps.count - 1].endTimestamp - steps[0].startTimestamp)/86400)
    var date = Date(timeIntervalSince1970: TimeInterval(steps[steps.count-1].endTimestamp))
    
    let dateFormatter = DateFormatter()
    dateFormatter.dateFormat = "dd-MM-yyyy"
    trip.endDate = dateFormatter.string(from: date);
    
    trip.masterImage = steps[steps.count-2].masterImage
    trip.isPublic = false;
    
    date = Date(timeIntervalSince1970: TimeInterval(steps[0].startTimestamp))
    trip.startDate = dateFormatter.string(from: date)
    trip.syncComplete = false;
    trip.temperature = steps[steps.count - 2].temperature;
  
    trip.tripName = PhotoLibraryProcessor.GenerateTripNameFromSteps(steps: steps)
    trip.distanceTravelled = steps[steps.count - 1].distanceTravelled;
    
    return trip;
  }
  
  static func GenerateTripNameFromSteps(steps: [StepModal]) -> String {
    var locations: [String] = ["", "Home"];
    var result: String = "";
    
    for step in steps{
      if let _ = locations.firstIndex(of: step.stepName) {
      
      } else {
        locations.append(step.stepName)
        result += step.stepName + ", ";
      }
      
      if locations.count > 4 {
        break;
      }
    }
    
    result = result.substring(to: result.lastIndex(of: ",") ?? String.Index(encodedOffset: result.count))
    return result;
  }
  
}

File: android/app/src/main/java/com/beerwithai/glimpse/Engine/Utils/PhotoLibraryProcessor.java
package com.beerwithai.glimpse.Engine.Utils;

import android.content.Context;
import android.database.Cursor;
import android.media.ExifInterface;
import android.provider.MediaStore;

import com.beerwithai.glimpse.Engine.Modals.ClusterModal;
import com.beerwithai.glimpse.Engine.Modals.HomesForDataClusteringModal;
import com.beerwithai.glimpse.Engine.Modals.Region;
import com.beerwithai.glimpse.Engine.Modals.StepModal;
import com.beerwithai.glimpse.Engine.Modals.TripModal;
import com.beerwithai.glimpse.Engine.Providers.DatabaseProvider;

import java.io.IOException;
import java.lang.reflect.Array;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;

import io.realm.RealmList;

public class PhotoLibraryProcessor {

    public static String GenerateTripNameFromSteps(ArrayList<StepModal> steps) {
        ArrayList<String> locations = new ArrayList<String>();
        locations.add("");
        locations.add("Home");

        String result = "";

        for(int i = 0; i < steps.size(); i++) {
            if(locations.indexOf(steps.get(i).stepName) != -1) { //TODO: Confirm if index is not found it's returning -1
                locations.add(steps.get(i).stepName);
                result += steps.get(i).stepName;
            }

            if(locations.size() > 4) {
                break;
            }
        }

        result = result.substring(0, result.lastIndexOf(','));
        return result;
    }

    public static TripModal PopulateTripWithSteps(TripModal trip, ArrayList<StepModal> steps) {
        trip.daysOfTravel = Integer.valueOf((int)(steps.get(steps.size() -1).endTimestamp - steps.get(0).startTimestamp)/86400);
        trip.startDate = ""; // TODO: Generate date from startTImestamp of steps
        trip.endDate = ""; // TODO: Generate date from endTimestamp of steps
        trip.masterImage = steps.get(steps.size() - 2).masterImage;
        trip.isPublic = false;
        trip.syncComplete = false;
        trip.temperature = steps.get(steps.size() - 2).temperature;
        trip.tripName = PhotoLibraryProcessor.GenerateTripNameFromSteps(steps);
        trip.distanceTravelled = steps.get(steps.size() - 1).distanceTravelled;
        trip.daysOfTravel = Integer.valueOf((int)(steps.get(steps.size() - 1).endTimestamp - steps.get(0).startTimestamp));
        return trip;
    }

    public static StepModal GetHomeStepFromTimestamp(ArrayList<HomesForDataClusteringModal> homesForDataClustering, double timestamp, String tripId, Integer stepId) {
        HomesForDataClusteringModal homeStep = homesForDataClustering.get((int)timestamp);

        StepModal _stepModal = new StepModal();
        _stepModal.stepName = "Home";
        _stepModal.stepId = stepId;
        _stepModal.tripId = tripId;
        _stepModal.meanLatitude = homeStep.latitude;
        _stepModal.meanLongitude = homeStep.longitude;
        _stepModal.startTimestamp = timestamp;
        _stepModal.endTimestamp = timestamp;

        Region markerRegion = new Region();
        markerRegion.latitude = homeStep.latitude;
        markerRegion.longitude = homeStep.longitude;

        _stepModal.markers = new RealmList<Region>();
        _stepModal.markers.add(markerRegion);

        return _stepModal;
    }

    public static TripModal PopulateTripModalData(ArrayList<StepModal> steps, String tripId, ArrayList<HomesForDataClusteringModal> homesForDataClustering) {
        TripModal tripResult = new TripModal();
        ArrayList<StepModal> _stepsForTrip = new ArrayList<StepModal>();
        ArrayList<String> countries = new ArrayList<String>();
        ArrayList<String> places = new ArrayList<String>();

        _stepsForTrip.add(PhotoLibraryProcessor.GetHomeStepFromTimestamp(homesForDataClustering, steps.get(0).startTimestamp, tripId, 0));

        for(int i = 1; i < _stepsForTrip.size(); i++) {
            try {
                String result = TripUtils.getLocationFromCoordinates(_stepsForTrip.get(i).meanLatitude, _stepsForTrip.get(i).meanLongitude);
                _stepsForTrip.get(i).stepName = result;
                _stepsForTrip.get(i).tripId = tripId;
                _stepsForTrip.get(i).stepId = i * 100;

                if (countries.indexOf(result) == -1) { // TODO: Check if indexOf gives -1 if not found
                    countries.add(result);
                    tripResult.countryCode.add(result);
                }

                if (places.indexOf(_stepsForTrip.get(i).stepName) == -1) {
                    places.add(_stepsForTrip.get(i).stepName);
                }

                _stepsForTrip.get(i).temperature = String.valueOf(TripUtils.getWeatherFromCoordinates(_stepsForTrip.get(i).meanLatitude, _stepsForTrip.get(i).meanLongitude));
                _stepsForTrip.get(i).distanceTravelled = steps.get(i - 1).GetDistanceBetween(_stepsForTrip.get(i));
            } catch (Exception e) {
                // TODO: Handle exception
            }
        }

        StepModal _stepModal2 = PhotoLibraryProcessor.GetHomeStepFromTimestamp(homesForDataClustering, steps.get(steps.size() - 1).endTimestamp, tripId, (steps.size() - 1)*100);
        _stepModal2.distanceTravelled = steps.get(steps.size() - 1).distanceTravelled + steps.get(steps.size() - 1).GetDistanceBetween(_stepModal2);
        _stepsForTrip.add(_stepModal2);

        tripResult.tripId = tripId;
        tripResult = PhotoLibraryProcessor.PopulateTripWithSteps(tripResult, _stepsForTrip);

        DatabaseProvider.UpdateDBWithSteps(_stepsForTrip);

        return tripResult;
    }

    public static ArrayList<TripModal> GenerateTripFromPhotos(ArrayList<ClusterModal> clusterData) throws Exception {
        ArrayList<HomesForDataClusteringModal> homesForDataClustering = DatabaseProvider.GetHomesForDataClustering();
        ArrayList<ArrayList<ClusterModal>> trips = ClusterProcessor.RunMasterClustering(clusterData, homesForDataClustering);

        if (trips.size() == 0) {
            throw new Exception("ABC"); //TODO: Write exceptions
        }

        ArrayList<TripModal> tripResult = new ArrayList<TripModal>();

        for(int i = 0; i < trips.size(); i++) {
            ArrayList<ClusterModal> _trip = trips.get(i);
            Collections.sort(_trip, new CompareClusterByTimestamp());
            ArrayList<StepModal> _unsortedSteps = ClusterProcessor.RunStepClustering(_trip);
            Collections.sort(_unsortedSteps, new CompareByTimestamp());
            TripModal trip = PhotoLibraryProcessor.PopulateTripModalData(_unsortedSteps, TripUtils.GenerateTripId(), homesForDataClustering);
            tripResult.add(trip);
        }
        Collections.sort(tripResult, new CompareTripsByTimestamp());
        return tripResult;
    }

    public static ArrayList<ClusterModal> getPhotosFromLibrary(Context context) {
        final Cursor cursor = context.getContentResolver().query(MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                null,
                null,
                null,
                null
        );

        ArrayList<ClusterModal> cameraImagePaths = new ArrayList<>();

        if (cursor == null) {
            return cameraImagePaths;
        }

        if (cursor.moveToLast()) {
            final int dataColumn = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA);
            do {
                String localPath = cursor.getString(dataColumn);
                ClusterModal _m = new ClusterModal();
                _m.image = localPath;
                cameraImagePaths.add(_m);
            } while (cursor.moveToPrevious());
        }
        cursor.close();

        return cameraImagePaths;
    }

    public static ArrayList<ClusterModal> PopulateImageProperties(ArrayList<ClusterModal> cluster) {

        for(ClusterModal c : cluster) {
            try {
                double timestamp = PhotoLibraryProcessor.computeTimestamp(c.image);
                Region region = PhotoLibraryProcessor.computeLocation(c.image);
                c.timestamp = timestamp;
                c.latitude = region.latitude;
                c.longitude = region.longitude;
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        return cluster;
    }

    private static double computeTimestamp(String localPath) throws Exception {
        ExifInterface exif = new ExifInterface(localPath);
        SimpleDateFormat fmt = new SimpleDateFormat("yyyy:MM:dd HH:mm:ss");
        String dateString = exif.getAttribute(ExifInterface.TAG_DATETIME);
        double timestamp = fmt.parse(dateString).getTime() / 1000;
        return timestamp;
    }

    private static Region computeLocation(String localPath) throws Exception {
        ExifInterface exif = new ExifInterface(localPath);
        SimpleDateFormat fmt = new SimpleDateFormat("yyyy:MM:dd HH:mm:ss");
        String longitude = exif.getAttribute(ExifInterface.TAG_GPS_LONGITUDE);
        String latitude = exif.getAttribute(ExifInterface.TAG_GPS_LATITUDE);
        Region _r = new Region();
        _r.latitude = Float.parseFloat(latitude);
        _r.longitude = Float.parseFloat(longitude);
        return _r;
    }

}

class CompareTripsByTimestamp implements Comparator<TripModal> {
    public int compare(TripModal a, TripModal b) {
        return a.startDate.compareTo(b.startDate);
    }
}
